### DISCUSSION TITLE
Date and Time Variables?

### DISCUSSION THREAD

**Discussion by mceh** [p2or/blender-loom#68] [Ideas]
Hi there,
thank you for your great work. Variables/Tokens in Render-names and file paths should be a native feature for years now. They are indispensable in other 3D packages like Cinema 4D. Thank you for plugging that hole.

BUT: Date and time variables are super important in any professional context and are needed all the time on bigger projects.

It’s quite easy to add that to  Loom. I did it myself. I think others would appreciate this too. This is what I did in the loom Python-file:

add to the imports at the top of the file, e.G. after line 26:
`import time`

after line 5422 add this to the global_var_defaults:
```
    "$YEAR": 'time.localtime()[0]',
    "$MONTH": 'time.localtime()[1]',
    "$DAY": 'time.localtime()[2]',
    "$HOUR": 'time.localtime()[3]',
    "$MINUTE": 'time.localtime()[4]',
    "$SECOND": 'time.localtime()[5]'
```
(don’t forget to add a comma to the preceding line 5422)

Remove and re-add the add-on. Reset the preferences in the add-on panel of Loom and you can use these new time variables.

Maybe the author will add this to the official download?!

Regards
Markus

**Comment by p2or** [comment]
Hi @mceh,

thanks for your suggestion. Since this is the [second time this request](https://github.com/p2or/blender-loom/issues/51) comes up and if this really helps to keep track of our work, I'd love to add them.

However, there is no need to modify the code of the add-on. The following expressions should work out of the box:

- `$YY`: `exec("import time") or time.strftime("%Y")`
- `$MM`: `exec("import time") or time.strftime("%m")`
- `$DD`: `exec("import time") or time.strftime("%d")`
- `$HH`: `exec("import time") or time.strftime("%H")`
- `$MI`: `exec("import time") or time.strftime("%M")`
- `$SD`: `exec("import time") or time.strftime("%S")`

Remaining question is, do we really need all this separately or wouldn't it make more sense to have one default variable to work with? E.g. `$TME`: `exec("import time") or time.strftime("%Y-%m-%d_%H-%M-%S")`:

<img width="800" alt="Screenshot 2023-10-23 at 09 50 44" src="https://github.com/p2or/blender-loom/assets/512368/1ee7ad25-f388-4e48-b0ce-838ff846be0c">


Cheers,
Christian

**Reply by mceh** [reply]
Hi @p2or ,
thanks for your reply. Sorry, I really didn’t find your suggested way to add the variables anywhere. Since date and time are so important, I’d suggest to add it to the Globals preset.

Regarding your question:
At least there should be separate variables for the date and the time. I worked on jobs where the workflow was that every day there was automatically generated a folder with the current date for all renders of the current day. In this folder where the renders with the exact time in the filename. This worked beautifully in Cinema 4D. It looked like:  //Renders/23-10-12/projectname-08_17_56.exr
A time variable combined with the date would not work for the automatic directory in this example.

Regards
Markus

**Comment by p2or** [comment]
Hi @mceh,

for dailies, you can easily split up the mentioned expression into (also quite legible IMHO):

 - `$DAY`: `exec("import time") or time.strftime("%Y-%m-%d")`
 - `$DAYTIME`: `exec("import time") or time.strftime("%H-%M-%S")`

In general, I would prefer to provide some default expressions for demo purposes that can be customized for any purpose instead of adding hundreds of them to support everything. Anyway, if there are other users who support this, we can gladly include both though. Cool?

In any case, this needs to be documented. If you have some time to help out I would greatly appreciate it.

Cheers,
Christian

**Comment by p2or** [comment]
Hi @mceh,

FYI: As of the latest commit (0d658ee) both expression are now part of *Loom* 🥳. If something else comes up in this context, just let me know. Closing for now, can be reopened at any time.

Cheers,
Christian
